ipn/ipnlocal/profiles.go

```diff
@@ -387,53 +511,94 @@ func (pm *profileManager) CurrentProfile() ipn.LoginProfile {
 	return *pm.currentProfile
 }
 
-// errProfileNotFound is returned by methods that accept a ProfileID.
+// errProfileNotFound is returned by methods that accept a ProfileID
+// when the specified profile does not exist.
 var errProfileNotFound = errors.New("profile not found")
 
+// errProfileAccessDenied is returned by methods that accept a ProfileID
+// when the current user does not have access to the specified profile.
+// It is used temporarily until we implement access checks based on the
+// caller's identity in tailscale/corp#18342.
+var errProfileAccessDenied = errors.New("profile access denied")
+
 // DeleteProfile removes the profile with the given id. It returns
-// errProfileNotFound if the profile does not exist.
+// [errProfileNotFound] if the profile does not exist, or an
+// [errProfileAccessDenied] if the specified profile is not accessible
+// to the current user.
 // If the profile is the current profile, it is the equivalent of
-// calling NewProfile() followed by DeleteProfile(id). This is
-// useful for deleting the last profile. In other cases, it is
-// recommended to call SwitchProfile() first.
+// calling [profileManager.NewProfile] followed by [profileManager.DeleteProfile](id).
+// This is useful for deleting the last profile. In other cases, it is
+// recommended to call [profileManager.SwitchProfile] first.
 func (pm *profileManager) DeleteProfile(id ipn.ProfileID) error {
 	metricDeleteProfile.Add(1)
-
-	if id == "" {
-		// Deleting the in-memory only new profile, just create a new one.
-		pm.NewProfile()
-		return nil
+	if id == pm.currentProfile.ID {
+		return pm.deleteCurrentProfile()
 	}
 	kp, ok := pm.knownProfiles[id]
 	if !ok {
 		return errProfileNotFound
 	}
-	if kp.ID == pm.currentProfile.ID {
+	if err := pm.checkProfileAccess(kp); err != nil {
+		return err
+	}
+	return pm.deleteProfileUnchecked(kp)
+}
+
+func (pm *profileManager) deleteCurrentProfile() error {
+	if err := pm.checkProfileAccess(pm.currentProfile); err != nil {
+		return err
+	}
+	if pm.currentProfile.ID == "" {
+		// Deleting the in-memory only new profile, just create a new one.
 		pm.NewProfile()
+		return nil
 	}
-	if err := pm.WriteState(kp.Key, nil); err != nil {
+	return pm.deleteProfileUnchecked(pm.currentProfile)
+}
+
+// deleteProfileUnchecked is like [profileManager.DeleteProfile],
+// but it doesn't check user's access rights to the profile.
+func (pm *profileManager) deleteProfileUnchecked(profile *ipn.LoginProfile) error {
```

Comment:
s/Unchecked/NoAccessCheck?

Reply:
or `NoPermCheck`?

Reply:
Went with `NoPermCheck` for this and other methods. Thanks!